### PR TITLE
Improve performance of printing JS ASTs

### PR DESCRIPTION
**Pull request by devongovett**
This replaces `@babel/generator` with [astring](https://github.com/davidbonnet/astring), another library for printing JS ASTs. Their benchmark claims it is up to 50x faster. In my own tests, it improved overall build time on the esbuild benchmark by 1.5x. In combination with #4366, this results in **~2.8x faster** dev builds. 🙌

Because Babel has its own AST format, and astring support only estree, I wrote a custom generator for astring that converts babel-specific nodes to estree on the fly. In addition, it deals with comments that can be attached to any node, whereas astring only support comments on statements.

My main question is around the quality of the source maps produced by astring as compared with babel generator but I'm not sure how to verify this. @DeMoorJasper?

### CODE REVIEW

**Review by DeMoorJasper — COMMENTED** [review]
Sourcemaps are apparently very broken with astring, attached an example from the kitchen sink app.

astring:
Screenshot 2020-12-06 at 18 38 12

Babel:
Screenshot 2020-12-06 at 18 38 27

File: packages/utils/astring/src/sourcemaps.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
+const VLQ_SHIFT = 5;
```
Comment (DeMoorJasper) [inline packages/utils/astring/src/sourcemaps.js:1]: This file isn't being used by astring, maybe we should just remove it?

Comment (devongovett) [reply packages/utils/astring/src/sourcemaps.js:1]: oh, weird

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>Kitchen Sink ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 4.02s | -207.00ms 🚀 |
| Cached | 783.00ms | +1.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.d5807e82.webp | 102.94kb | +0.00b | 48.00ms | +7.00ms ⚠️ |
| dist/modern/parcel.d5807e82.webp | 102.94kb | +0.00b | 100.00ms | +28.00ms ⚠️ |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.d5807e82.webp | 102.94kb | +0.00b | 54.00ms | -37.00ms 🚀 |
| dist/modern/parcel.d5807e82.webp | 102.94kb | +0.00b | 70.00ms | +11.00ms ⚠️ |
| dist/legacy/index.5acd07fb.js | 1.07kb | +0.00b | 76.00ms | -15.00ms 🚀 |
| dist/modern/index.110a671b.js | 1.07kb | +0.00b | 73.00ms | -18.00ms 🚀 |
| dist/legacy/index.html | 701.00b | +0.00b | 76.00ms | -15.00ms 🚀 |
| dist/modern/index.html | 701.00b | +0.00b | 73.00ms | -18.00ms 🚀 |
| dist/legacy/index.5859ab37.css | 77.00b | +0.00b | 74.00ms | -17.00ms 🚀 |
| dist/modern/index.77ce15d4.css | 77.00b | +0.00b | 73.00ms | -18.00ms 🚀 |

</p></details>

<details><summary>React HackerNews ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 22.89s | -229.00ms |
| Cached | 13.80s | -2.31s 🚀 |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/PermalinkedComment.52a3fc6f.js | 4.22kb | +0.00b | 64.00ms | -4.00ms 🚀 |
| dist/UserProfile.42c44c22.js | 1.70kb | +0.00b | 64.00ms | -4.00ms 🚀 |
| dist/logo.24c8bf9e.png | 274.00b | +0.00b | 31.00ms | -2.00ms 🚀 |

</p></details>

<details><summary>AtlasKit Editor ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 2.18m | +5.55s |
| Cached | 2.97s | +1.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.152dbfb8.js | 2.34mb | +0.00b | 1.34m | +5.38s ⚠️ |
| dist/pdfRenderer.f8527c1c.js | 1.11mb | +0.00b | 41.96s | -4.41s 🚀 |
| dist/editorView.4de8e34d.js | 587.39kb | +0.00b | 1.05m | -5.30s 🚀 |
| dist/popup.c7d523d4.js | 169.35kb | +0.00b | 1.34m | +1.06m ⚠️ |
| dist/EmojiPickerComponent.13c2889c.js | 139.08kb | +0.00b | 10.13s | -540.00ms 🚀 |
| dist/Toolbar.86dd17ff.js | 99.25kb | +0.00b | 43.40s | -6.11s 🚀 |
| dist/card.f3741dfa.js | 53.76kb | +0.00b | 11.40s | -1.06m 🚀 |
| dist/card.2fa3ec33.js | 51.51kb | +0.00b | 14.26s | -32.78s 🚀 |
| dist/Modal.34634fc9.js | 39.03kb | +0.00b | 3.66s | -1.31s 🚀 |
| dist/component.088d2ae3.js | 30.87kb | +0.00b | 3.99s | +2.03s ⚠️ |
| dist/component.22035ce4.js | 22.40kb | +0.00b | 3.65s | -1.31s 🚀 |
| dist/smartMediaEditor.cfe1bdad.js | 16.79kb | +0.00b | 40.74s | -7.50s 🚀 |
| dist/js.0a808d22.js | 16.55kb | +0.00b | 3.65s | -1.31s 🚀 |
| dist/dropzone.374e8cb6.js | 16.03kb | +0.00b | 14.24s | -11.54s 🚀 |
| dist/ui.078d3b2e.js | 14.11kb | +0.00b | 11.34s | -982.00ms 🚀 |
| dist/workerHasher.83104f73.js | 11.90kb | +0.00b | 7.30s | -1.43s 🚀 |
| dist/component.1880fd07.js | 6.27kb | +0.00b | 4.03s | -1.59s 🚀 |
| dist/media-viewer.083ce2a1.js | 4.01kb | +0.00b | 12.97s | -2.15s 🚀 |
| dist/EmojiPickerComponent.9a5c1b76.js | 3.68kb | +0.00b | 7.40s | -29.89s 🚀 |
| dist/png-chunks-extract.692b85d1.js | 3.55kb | +0.00b | 3.65s | -1.31s 🚀 |
| dist/index.10f02a5e.css | 3.46kb | +0.00b | 44.34s | -6.61s 🚀 |
| dist/dropzone.c5e4485d.js | 3.44kb | +0.00b | 1.34m | +54.86s ⚠️ |
| dist/Modal.420db102.js | 3.15kb | +0.00b | 1.07s | -1.46s 🚀 |
| dist/clipboard.a67b66af.js | 2.97kb | +0.00b | 14.25s | -32.13s 🚀 |
| dist/16.93e3596b.js | 2.49kb | +0.00b | 1.07s | -876.00ms 🚀 |
| dist/ResourcedEmojiComponent.557284bd.js | 2.15kb | +0.00b | 7.40s | -4.91s 🚀 |
| dist/card.09f4d8e9.js | 2.15kb | +0.00b | 1.32m | +1.07m ⚠️ |
| dist/date.c50d5e8c.js | 1.96kb | +0.00b | 4.44s | -1.12s 🚀 |
| dist/images.c0aa934a.js | 1.90kb | +0.00b | 4.44s | -1.12s 🚀 |
| dist/feedback.0e55f851.js | 1.86kb | +0.00b | 4.87s | -3.75s 🚀 |
| dist/16.8ab9c9cf.js | 1.86kb | +0.00b | 4.21s | -1.11s 🚀 |
| dist/browser.dfefe24a.js | 1.82kb | +0.00b | 14.26s | -32.78s 🚀 |
| dist/16.aa2b1e8c.js | 1.79kb | +0.00b | 4.03s | -1.74s 🚀 |
| dist/workerHasher.245b987a.js | 1.75kb | +0.00b | 10.11s | +1.39s ⚠️ |
| dist/workerHasher.f5a304ed.js | 1.75kb | +0.00b | 22.34s | -52.93s 🚀 |
| dist/list-number.41f28a70.js | 1.68kb | +0.00b | 4.57s | -1.10s 🚀 |
| dist/status.eb15cda1.js | 1.68kb | +0.00b | 4.77s | -1.01s 🚀 |
| dist/code.d63c0558.js | 1.61kb | +0.00b | 4.33s | -1.12s 🚀 |
| dist/link.a9025e77.js | 1.53kb | +0.00b | 4.57s | -1.10s 🚀 |
| dist/heading6.053111a7.js | 1.53kb | +0.00b | 4.87s | -1.03s 🚀 |
| dist/heading3.9115aad6.js | 1.51kb | +0.00b | 4.87s | -1.03s 🚀 |
| dist/16.15d9f682.js | 1.46kb | +0.00b | 4.03s | -1.59s 🚀 |
| dist/emoji.d5ab3e45.js | 1.45kb | +0.00b | 4.44s | -1.13s 🚀 |
| dist/16.7fc8af6d.js | 1.45kb | +0.00b | 1.07s | -882.00ms 🚀 |
| dist/16.474c8276.js | 1.44kb | +0.00b | 4.33s | -1.11s 🚀 |
| dist/16.d49a0bb7.js | 1.40kb | +0.00b | 4.33s | -1.11s 🚀 |
| dist/heading5.731efbe4.js | 1.40kb | +0.00b | 4.87s | -1.03s 🚀 |
| dist/expand.7a2a1b06.js | 1.38kb | +0.00b | 10.09s | +1.47s ⚠️ |
| dist/16.c4941d0d.js | 1.36kb | +0.00b | 4.21s | -1.11s 🚀 |
| dist/16.a4fa1a69.js | 1.34kb | +0.00b | 1.07s | -875.00ms 🚀 |
| dist/heading2.de347970.js | 1.33kb | +0.00b | 4.77s | -1.13s 🚀 |
| dist/16.ad29438e.js | 1.32kb | +0.00b | 4.21s | -1.11s 🚀 |
| dist/16.3d7a389d.js | 1.31kb | +0.00b | 4.33s | -1.12s 🚀 |
| dist/media-card-analytics-error-boundary.5ce95433.js | 1.30kb | +0.00b | 11.36s | -1.07m 🚀 |
| dist/mention.1703453a.js | 1.29kb | +0.00b | 4.57s | -1.10s 🚀 |
| dist/heading4.5089aec7.js | 1.29kb | +0.00b | 4.87s | -1.03s 🚀 |
| dist/Modal.ae6f5f7a.js | 1.28kb | +0.00b | 3.65s | -1.31s 🚀 |
| dist/16.55b14beb.js | 1.27kb | +0.00b | 1.07s | -879.00ms 🚀 |
| dist/layout.32cc11b1.js | 1.27kb | +0.00b | 4.57s | -1.10s 🚀 |
| dist/16.45835899.js | 1.27kb | +0.00b | 4.21s | -1.11s 🚀 |
| dist/16.e83a0923.js | 1.26kb | +0.00b | 4.05s | -1.10s 🚀 |
| dist/16.71387abf.js | 1.26kb | +0.00b | 4.03s | -1.59s 🚀 |
| dist/divider.6cc4f21a.js | 1.25kb | +0.00b | 4.44s | -1.13s 🚀 |
| dist/quote.b905d023.js | 1.25kb | +0.00b | 4.77s | -1.01s 🚀 |
| dist/component.242a45f9.js | 1.23kb | +0.00b | 3.95s | +1.43s ⚠️ |
| dist/action.3a733320.js | 1.23kb | +0.00b | 4.33s | -1.12s 🚀 |
| dist/16.f1c68cda.js | 1.23kb | +0.00b | 3.95s | +1.42s ⚠️ |
| dist/decision.1f5dd6ad.js | 1.21kb | +0.00b | 4.44s | -1.12s 🚀 |
| dist/panel-warning.7dcb372a.js | 1.21kb | +0.00b | 4.68s | -1.10s 🚀 |
| dist/16.8cd85281.js | 1.18kb | +0.00b | 4.21s | -1.11s 🚀 |
| dist/list.7d80b06a.js | 1.18kb | +0.00b | 4.57s | -1.10s 🚀 |
| dist/heading1.c1474f93.js | 1.18kb | +0.00b | 4.77s | -2.20s 🚀 |
| dist/panel-error.4dcbeea1.js | 1.11kb | +0.00b | 4.68s | -2.09s 🚀 |
| dist/panel.9407e91b.js | 1.10kb | +0.00b | 4.68s | -1.10s 🚀 |
| dist/table.547bc9c1.js | 1.09kb | +0.00b | 4.77s | -1.01s 🚀 |
| dist/media-picker-analytics-error-boundary.607a1e52.js | 1.05kb | +0.00b | 1.34m | +54.88s ⚠️ |
| dist/panel-success.23208184.js | 1.05kb | +0.00b | 4.68s | -2.09s 🚀 |
| dist/media-card-analytics-error-boundary.f92fc998.js | 1.05kb | +0.00b | 11.35s | -1.07m 🚀 |
| dist/media-card-analytics-error-boundary.36e013e9.js | 1.05kb | +0.00b | 22.34s | -24.71s 🚀 |
| dist/panel-note.7a985368.js | 1.05kb | +0.00b | 4.68s | -2.09s 🚀 |
| dist/media-viewer-analytics-error-boundary.cceb3bd6.js | 995.00b | +0.00b | 12.97s | -2.15s 🚀 |
| dist/simpleHasher.26a74474.js | 755.00b | +0.00b | 7.28s | -1.44s 🚀 |
| dist/simpleHasher.042b95c1.js | 755.00b | +0.00b | 40.63s | -7.60s 🚀 |
| dist/index.html | 119.00b | +0.00b | 3.99s | +823.00ms ⚠️ |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/pdfRenderer.f8527c1c.js | 1.11mb | +3.00b ⚠️ | 674.00ms | +11.00ms |
| dist/png-chunks-extract.692b85d1.js | 3.55kb | +0.00b | 589.00ms | +84.00ms ⚠️ |
| dist/16.fe7e09db.js | 1.46kb | +0.00b | 542.00ms | -38.00ms 🚀 |

</p></details>

<details><summary>Three.js ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 15.41s | +30.00ms |
| Cached | 720.00ms | +15.00ms |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

*No bundle changes detected.*

</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/286534073584714248)**

File: packages/shared/babel-ast-utils/src/generator.js [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+    state.write(';');
+  },
+  PrivateName(node, state) {
+    state.write('#' + node.name, node);
+  },
+  Import(node, state) {
+    // astring doesn't support ImportExpression yet
+    state.write('import');
+  },
+  // TODO: OptionalMemberExpression, OptionalCallExpression once astring supports ChainExpression
```
Comment (DeMoorJasper) [inline packages/shared/babel-ast-utils/src/generator.js:158]: We can probably put this in our fork as well?

https://github.com/davidbonnet/astring/pull/424

Not sure if the test in the PR shouldn't actually just be a fixture instead, seems easier.

**Review by DeMoorJasper — COMMENTED** [review]
Looks good to me although not sure if it will be as stable as Babel, maybe we should try it on some larger codebases?

Also maybe we should add this patch as well to our fork: https://github.com/davidbonnet/astring/pull/315

File: packages/shared/babel-ast-utils/src/generator.js (hunk 2 above)
Comment (devongovett) [reply packages/shared/babel-ast-utils/src/generator.js]: done!

**Review by wbinnssmith — APPROVED** [review]

**Pull request merged**